packages/zoe/src/zoeService/zoe.js

```diff
@@ -44,37 +43,80 @@ import { ZoeServiceIKit } from '../typeGuards.js';
  * @param {Baggage} [zoeBaggage]
  */
 const makeZoeKit = (
-  vatAdminSvc,
+  vatAdminSvcP = undefined,
   shutdownZoeVat = () => {},
   feeIssuerConfig = defaultFeeIssuerConfig,
   zcfSpec = { name: 'zcf' },
   zoeBaggage = makeScalarBigMapStore('zoe baggage', { durable: true }),
 ) => {
-  let zoeService;
+  let zcfBundleCap;
+
+  function saveBundleCap() {
+    E.when(
+      Promise.all([vatAdminSvcP, getZcfBundleCap(zcfSpec, vatAdminSvcP)]),
+      ([vatAdminService, bundleCap]) => {
+        zcfBundleCap = bundleCap;
+
+        zoeBaggage.init('vatAdminSvc', vatAdminService);
+        zoeBaggage.init('zcfBundleCap', zcfBundleCap);
+      },
+    );
+  }
+
+  const setVatAdminService = lateVatAdminSvc => {
+    vatAdminSvcP = lateVatAdminSvc;
+    saveBundleCap();
+  };
+  if (vatAdminSvcP) {
+    saveBundleCap();
+  } else if (zoeBaggage.has('zcfBundleCap')) {
+    zcfBundleCap = zoeBaggage.get('zcfBundleCap');
+    // in this case, it'll be known to have been resolved, but that's fine
+    vatAdminSvcP = zoeBaggage.get('vatAdminSvc');
+  }
 
   const feeMintKit = vivifyFeeMint(zoeBaggage, feeIssuerConfig, shutdownZoeVat);
 
   /** @type {GetBundleCapForID} */
-  const getBundleCapForID = bundleID =>
-    E(vatAdminSvc).waitForBundleCap(bundleID);
+  const getBundleCapForID = bundleID => {
+    vatAdminSvcP || Fail`vatAdminSvc must be defined.`;
+    // @ts-expect-error the guard protects it.
```

Comment:
Should these checks surface the inner `assert`?
```suggestion
    assert(vatAdminSvcP, 'vatAdminSvc must be defined.');
```
or alternatively update the remaining `assert(expression, 'static message')` guards to ``` expression || Fail`static message` ```?

Reply:
@gibson042 we're trying to convert all occurrences of something like 
```js
assert(vatAdminSvcP, 'vatAdminSvc must be defined.');
```
to
```js
 vatAdminSvcP || Fail`vatAdminSvc must be defined.`;
```
Your recommendation seems to reverse that. If the problem is that it confuses TS, there are two workarounds that I prefer to returning to the old style. The `@ts-expect-error` comment already in there, or better
```suggestion
    if (!vatAdminSvcP) {
      throw Fail`vatAdminSvc must be defined.`;
    }
```
The `throw` is redundant since `Fail` itself always throws. It is typed as returning `never` which should give TS enough info. But TS sometimes gets confused about this anyway, and the redundant `throw` unconfuses it.

Believe it of not, this is all actually documented ;)
https://github.com/Agoric/agoric-sdk/blob/19855de77cca7a96c1c3a42f090502d11ba36989/packages/assert/src/types.js#L282-L338

Reply:
Excellent, that's perfect. Suggestion updated.

Reply:
done